dali/python/nvidia/dali/external_source.py

```diff
@@ -420,23 +420,39 @@ class ExternalSource():
 
     Acceptable sources depend on the value specified for ``batch`` parameter.
 
-    If batch is set to False, source must be a callable (a function or an object with __call__ method)
-    that accepts exactly one argument (:meth:`~nvidia.dali.types.SampleInfo` instance that
-    represents the index of the requested sample).
-    If batch is set to True, the ``source`` can be either a callable, an iterable or a generator function.
-    Callable in batch mode must accept exactly one argument - either :meth:`~nvidia.dali.types.BatchInfo`
-    instance or an integer (see `batch_info`).
+    If ``batch`` is set to ``False``, the ``source`` must be:
 
-    Irrespective of ``batch`` value, callables should produce requested sample or batch solely based on
-    the SampleInfo/BatchInfo instance or index in batch, so that they can be run in parallel in a number of workers.
-    When batch is set to True, callables performance might especially benefit from increasing
-    ``prefetch_queue_depth`` so that a few next batches can be computed in parallel.
+      * a callable (a function or an object with __call__ method) that accepts exactly one argument
+        (:class:`~nvidia.dali.types.SampleInfo` instance that represents the index of the requested sample).
 
-    Iterator or generator will be assigned to a single worker that will iterate over it.
+    If ``batch`` is set to ``True``, the ``source`` can be either:
 
-    The ``source`` callback must raise StopIteration when the end of data is reached. Note, that due to
-    prefetching, the callback may be invoked with a few iterations past the end of dataset - make sure
-    it consistently raises StopIteration in that case.
+      * a callable that accepts exactly one argument (either :class:`~nvidia.dali.types.BatchInfo`
+        instance or an integer - see ``batch_info`` for details)
+      * an iterable,
+      * a generator function.
+
+    .. warning::
+        Irrespective of ``batch`` value, callables should produce requested sample or batch solely
+        based on the :class:`~nvidia.dali.types.SampleInfo`/:class:`~nvidia.dali.types.BatchInfo`
+        instance or index in batch, so that they can be run in parallel in a number of workers.
```

Comment:
Is this about not assuming any particular ordering? If so, I'd add a parenthesis with such clarification

Reply:
Not only, it's also about being stateless. I'm not sure how to clarify it better.

Reply:
Added mention of being stateless